Option to render edges

aothms opened the pull request:
* Option to render edges
* Sync to pythreejs 1.0.0b3

This is not really intended to be merged, just a quick fyi on my progress. It seems that backwards compatibility with the 0.4 branch would require a substantial amount of extra code (not just the couple of lines I added), so we need to see what version(s) you are planning to target in the end.

![notebook_edges](https://user-images.githubusercontent.com/1096535/34382532-e0bc8db4-eb0f-11e7-840d-a93016971790.png)

Things I am still planning to work on:
* Navigation and better initial camera rotation center point
* Interactivity (clickable surface and callback fn)
* Default jupyter repr function on TopoDS_Shape

aothms commented on src/Display/WebGl/jupyter_renderer.py:
```diff
@@ -87,35 +98,59 @@ def DisplayShape(self, shp, shape_color = default_shape_color):
         shape_mesh = Mesh(geometry=shape_geometry, material = shp_material)
-        scene_shp = Scene(children=[shape_mesh, AmbientLight(color='#101010')])
 
+        edges = list(map(lambda i_edge: [tess.GetEdgeVertex(i_edge, i_vert) for i_vert in range(tess.ObjEdgeGetVertexCount(i_edge))], range(tess.ObjGetEdgeCount())))
```
Oops, this still needs to go under the `if render_edges:` block

tpaviot commented:
Excellent. Backward compatibility with 0.4 is not a requirement, I will drop 0.4 support as soon as I get 1.0.0b3 working ! You can thus remove all 0.4 only code.

By the way, did you do something special to install pythreejs1.0.0b3 ? "special" in  the sense "not documented on the pythreejs page"

tpaviot commented:
@aothms I confirm that the target is pythreejs 1.0.0.bxxx.

aothms commented:
> By the way, did you do something special to install pythreejs1.0.0b3 ? "special" in the sense "not documented on the pythreejs page"

I am working on this on a jupyterhub docker spawner ubuntu image. In case it might help you can find the relevant part below. It took me some fiddling with users and paths [1].

[1] https://github.com/jovyan/pythreejs/pull/150

~~~dockerfile
WORKDIR /opt/build
RUN git clone https://github.com/jovyan/pythreejs
WORKDIR /opt/build/pythreejs
RUN chown -R jovyan .
USER jovyan
RUN /opt/conda/bin/pip install --user -e .
WORKDIR /opt/build/pythreejs/js
RUN npm run autogen
RUN npm run build:all
USER root
RUN jupyter nbextension install --py --symlink --sys-prefix pythreejs
RUN jupyter nbextension enable pythreejs --py --sys-prefix
~~~

Unfortunately I have not yet succeeded in making selection work [2] (which was one of the main use cases for me for migrating to pythreejs) so that's a bit of a bummer.

[2] https://github.com/jovyan/pythreejs/issues/152

I have worked a bit on multiple shapes in the same viewport, I added kwarg `show` and `Display()` method for that. What do you think? Perhaps `display(j)` would also be an option, `AddShape()`?

~~~py
j = jupyter_renderer.JupyterRenderer()
j.DisplayShape(box1, render_edges=True, show=False)
j.DisplayShape(box2, render_edges=True, shape_color=(1.,0.,0.), edge_color=(0.,1.,0.), show=False)
j.Display()
~~~

tpaviot commented:
Selection work: I use to work with 3js and BufferGeometries for a while, selection works perfectly. Run the core_webgl_threejs_* examples from the latest pythonocc-0.18.1 release, and check the related code.

Attached a zip archive including all files generated from the torus examples. You have a basic selection/event implemented.
[toruses_pyocc0181_3jsr88.zip](https://github.com/tpaviot/pythonocc-core/files/1591633/toruses_pyocc0181_3jsr88.zip)

The selection mechanism also required to build a mapping between a buffergeometry Id and the related topods_shape.

Adding shapes: to be consistent with GUI based renderer, something like this would be cool:

renderer.DisplayShape(shape1)
renderer.DisplayShape(shape2)
renderer.start_display()

shape1 and shape2 would be added to a shapelist to be renderer.

tpaviot commented:
I finally scceeded to install the current dev version. I studied the conda-forge recipe and extracted the following line:
```python setup.py install --single-version-externally-managed --record record.txt ```
and then install/enable the jupyter extension.
Tested on my windows 10 machine in a dedicated conda environment

tpaviot commented:
I merged your first commit (option to render edges, amended see 9104870bab31d3f5b65e6ab626e6bc9d4c4f1b91) and bump to 1.0.0x version. It works as expected. Moving up to pythreejs 1.0.0b3, I noticed a few regressions:

* the 3d window is much smaller. I don't know how to set the widget initial size ;

* the quality seems to be worse (a lot of aliasing) ;

* the zoom, using the middle mouse rolling button, does not work as expected.

did you ?

aothms commented:
> I finally scceeded to install the current dev version

Wonderful, I removed the few backwards comp lines

> ... and extracted the following line:

Magic at work

> the 3d window is much smaller. I don't know how to set the widget initial size

> the quality seems to be worse (a lot of aliasing) ;

I have a width, height on the Renderer (https://github.com/tpaviot/pythonocc-core/pull/486/commits/46253d84a8f7136e572f5ab42545ae2514dc29d3#diff-b509ee812727b2a79c36a29c0479574fR255). Maybe that's different from your approach of size (perhaps that simply scales the pixels but leaves the opengl viewport the same, I didn't have a look).

> the zoom, using the middle mouse rolling button, does not work as expected.

I see we arrived at the same conclusion and used OrbitControls :)

---

I have selection in a reasonably working state now, but you need my PR in pythreejs as well: https://github.com/jovyan/pythreejs/pull/153 at the very least the first commit to deal with missing UVs. In my code the edge linesegments are no longer candidates for picking so the 2nd commit there is not strictly necessary.

aothms commented:
There is also the option of `antialias = True` on the `Renderer` btw.

tpaviot commented:
I also noticed the width, height and antialias parameters. I picked-up your first commit (render edges), and commited a patch over it. you can rebase your work on top of review/tess and submit another PR

tpaviot commented:
@aothms The latest commit of review/tess is sync with your progresses, thank you. I modified a bit the selection mechanism. This works quite well with the changes you pushed to the pythreejs repos regarding mesh picking.

aothms commented:
Cool, thanks for getting things in sync

tpaviot commented:
I close this PR, everything has been merged into review/tess (will soon merge into master). Please submit another PR if you make progresses.
